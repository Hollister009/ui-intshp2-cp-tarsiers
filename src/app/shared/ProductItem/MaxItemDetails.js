/* eslint-disable no-console */
import React, { Component } from 'react';
import { Flags } from 'react-feature-flags';
import { Notify } from 'react-redux-notify';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

import HttpService from '../../../utils/http.service';
import appConfig from '../../../config/appConfig';
import productType from '../../../types';
import NotifyService from '../../../utils/notify.service';

const { addToWishList, removeFromWishList } = appConfig.apiResources;

const CN = 'product-item--full';

class MaxItemDetails extends Component {
  static propTypes = {
    /**
     * data - productType shape
     */
    data: productType,
    addToWishListItem: PropTypes.func.isRequired,
    removeFromWishListItem: PropTypes.func.isRequired,
    wished: PropTypes.bool
  };

  static defaultProps = { data: null, wished: false };

  state = { heartDisabled: false };

  componentDidMount() {
    const { data } = this.props;

    this.setState({ image: data.src });
  }

  addItem = id => {
    const { addToWishListItem, wished, createNotification } = this.props;

    HttpService.post(addToWishList, { productId: id })
      .then(res => {
        if (res.status === 200 && !wished) {
          addToWishListItem(id);
          console.log(`Added to the WishList: ${id}`);
        }
      })
      .catch(error => console.log(error))
      .finally(this.setState({ heartDisabled: false }));

    createNotification(NotifyService.added);
  };

  removeItem = id => {
    const { removeFromWishListItem, createNotification } = this.props;

    HttpService.post(removeFromWishList, { productId: id })
      .then(res => {
        if (res.status === 200) {
          removeFromWishListItem(id);
          console.log(`Removed from the WishList: ${id}`);
        }
      })
      .catch(error => console.log(error))
      .finally(this.setState({ heartDisabled: false }));

    createNotification(NotifyService.removed);
  };

  toggleSwatch = (e, colors) => {
    if (e.keyCode === 13 || e.type === 'click') {
      const value = e.target.innerText;

      this.setState({ image: colors[value] });
    }
  };

  swatchesFactory = (pins, items) =>
    pins.map(color => (
      <span
        key={color}
        role="button"
        tabIndex="0"
        style={{ backgroundColor: `${color}` }}
        onClick={e => this.toggleSwatch(e, items)}
        onKeyDown={e => this.toggleSwatch(e, items)}
      >
        {color}
      </span>
    ));

  toggleWishList = (e, id) => {
    const { wished } = this.props;

    e.preventDefault();
    const cb = !wished ? this.addItem : this.removeItem;

    this.setState({ heartDisabled: true }, () => {
      cb(id);
    });
  };

  cartNote = () => {
    const { createNotification } = this.props;

    createNotification(NotifyService.cart);
  };

  render() {
    const { data } = this.props;
    const { _id, title, sizes, colors, colorUrls, wished } = data;
    const { heartDisabled, image } = this.state;
    const allSizes = sizes
      .map((size, i) => (i !== 0 ? `- ${size}` : size))
      .join(' ');
    const swatches = this.swatchesFactory(colors, colorUrls);

    return (
      data && (
        <React.Fragment>
          <Link to={`/products/${_id}`}>
            <img className={`${CN}__img-small`} src={image} alt={data.title} />
            <h4 className="highlighted">{title}</h4>
          </Link>
          <div className={`${CN}__sizes`}>{`sizes : ${allSizes}`}</div>
          <div className={`${CN}__swatches`}>{swatches}</div>
          <hr className="separate" />
          <div className="social_buttons">
            <button type="button" title="Share with others">
              <i className="fas fa-share-alt" />
            </button>
            <Notify position={NotifyService.position.topRight} />
            <button
              type="button"
              title="Add to shopping-cart"
              onClick={this.cartNote}
            >
              <i className="fas fa-shopping-cart" />
            </button>
            <Flags authorizedFlags={[appConfig.killswitch.wishlist]}>
              <button
                type="button"
                className="btn-heart"
                onClick={e => this.toggleWishList(e, _id)}
                title="Add to wish-list"
                disabled={heartDisabled}
              >
                <i
                  className={
                    wished ? 'fas fa-heart highlighted' : 'fas fa-heart'
                  }
                />
              </button>
            </Flags>
          </div>
        </React.Fragment>
      )
    );
  }
}

export default MaxItemDetails;
