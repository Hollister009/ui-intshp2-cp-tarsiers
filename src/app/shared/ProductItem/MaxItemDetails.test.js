import React from 'react';
import { shallow } from 'enzyme';
import MaxItemDetails from './MaxItemDetails';

jest.mock('../../actions');
const products = require('../../../mocks/products.json');

describe('<MaxItemDetails />', () => {
  let props;

  beforeEach(() => {
    props = {
      data: products[0],
      addToWishListItem: jest.fn(),
      removeFromWishListItem: jest.fn(),
      wished: true
    };
  });

  it('should match its snapshot', () => {
    const wrapper = shallow(<MaxItemDetails {...props} />);

    expect(wrapper).toMatchSnapshot();
  });
});
