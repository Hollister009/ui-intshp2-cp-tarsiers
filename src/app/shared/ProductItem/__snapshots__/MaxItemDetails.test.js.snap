// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MaxItemDetails /> should match its snapshot 1`] = `
<Fragment>
  <img
    alt=""
    className="product-item--full__img-small"
    src="https://res.cloudinary.com/so/image/upload/v1548069002/products/reebok%20jacket/red.jpg"
  />
  <h4
    className="highlighted"
  >
    Reebok Track Jacket
  </h4>
  <div
    className="product-item--full__sizes"
  >
    sizes : s - m - l - xl
  </div>
  <div
    className="product-item--full__swatches"
  >
    <span
      key="#e63a43"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      style={
        Object {
          "backgroundColor": "#e63a43",
        }
      }
      tabIndex="0"
    >
      #e63a43
    </span>
    <span
      key="#5556ef"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      style={
        Object {
          "backgroundColor": "#5556ef",
        }
      }
      tabIndex="0"
    >
      #5556ef
    </span>
    <span
      key="#48d250"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      style={
        Object {
          "backgroundColor": "#48d250",
        }
      }
      tabIndex="0"
    >
      #48d250
    </span>
    <span
      key="#a04ad6"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      style={
        Object {
          "backgroundColor": "#a04ad6",
        }
      }
      tabIndex="0"
    >
      #a04ad6
    </span>
  </div>
  <hr
    className="separate"
  />
  <div
    className="social_buttons"
  >
    <button
      title="Share with others"
      type="button"
    >
      <i
        className="fas fa-share-alt"
      />
    </button>
    <button
      title="Add to shopping-cart"
      type="button"
    >
      <i
        className="fas fa-shopping-cart"
      />
    </button>
    <Flags
      authorizedFlags={
        Array [
          "showWishlist",
        ]
      }
      exactFlags={false}
      renderOff={[Function]}
      renderOn={[Function]}
    >
      <button
        className="btn-heart"
        disabled={false}
        onClick={[Function]}
        title="Add to wish-list"
        type="button"
      >
        <i
          className="fas fa-heart"
        />
      </button>
    </Flags>
  </div>
</Fragment>
`;
